### DISCUSSION TITLE
How do you invoke a lambda inside another lambda using Micronaut

### DISCUSSION THREAD

**Discussion by luiscm** [micronaut-projects/micronaut-core#5724] [General]
Dear friends

Because using AWS API is done like this below:
I would like if someone can help me with this detail using Micronaut.
I would like a helloworld project about this
 
 
   public static final String LAMBDA_FUNCTION_NAME = "lambdaToBeInvoked";
   public static final String DEFAULT_LAMBDA_REGION = "us-east-1";

   private void invokeLambda(String keyName, String bucketName) throws JSONException {

        JSONObject payloadObject = new JSONObject();
        payloadObject.put("keyName", keyName);
        payloadObject.put("bucketName", bucketName);

        String payload = payloadObject.toString();

        logger.info("Payload: " + payload);
        logger.info("Invoking Lambda: " + LAMBDA_FUNCTION_NAME);

        AWSLambda client = AWSLambdaAsyncClient.builder().withRegion(DEFAULT_LAMBDA_REGION).build();

        InvokeRequest request = new InvokeRequest();
        request.withFunctionName(LAMBDA_FUNCTION_NAME).withPayload(payload);
        InvokeResult invoke = client.invoke(request);
        logger.info("Result invoking " + LAMBDA_FUNCTION_NAME + ": " + invoke);
    }

**Comment by jeffscottbrown** [comment]
@luiscm Does that code do what you expect it to do?  

Are you looking for something in the Micronaut framework that will simplify that?

Thanks for any clarification.

**Reply by luiscm** [reply]
That's right, I want to know how to simplify this using Micronaut.
If you have a hello world for that, it would be a good example to have on the Micronaut website.

**Reply by jeffscottbrown** [reply]
We do have support for invoking AWS functions although it seems the documentation was lost when the function aws module was extracted out of core. See [function-client-aws/src/test/groovy/io/micronaut/function/client/aws/AwsLambdaInvokeSpec.groovy](https://github.com/micronaut-projects/micronaut-aws/blob/4c8073117bf18ee1a1e0004101f537ebd8171d49/function-client-aws/src/test/groovy/io/micronaut/function/client/aws/AwsLambdaInvokeSpec.groovy).

**Reply by luiscm** [reply]
I want a hello world in Java.
Do you know if this is possible

**Reply by jeffscottbrown** [reply]
There are examples of using @FunctionClient at [docs.micronaut.io/1.3.7/guide/index.html#callingFunctions](https://docs.micronaut.io/1.3.7/guide/index.html#callingFunctions).

**Reply by luiscm** [reply]
I just wanted a demo project in Java.

**Reply by jeffscottbrown** [reply]
I will create an example that renders "hello world" and will upload it for you.

**Reply by luiscm** [reply]
Thanks, wait...

**Reply by jeffscottbrown** [reply]
See the project at [github.com/jeffbrown/luiscmfunctionclient](https://github.com/jeffbrown/luiscmfunctionclient).

[src/main/resources/application.yml](https://github.com/jeffbrown/luiscmfunctionclient/blob/972346efe5b3fc8ef3f63e0f3331b522410c6966/src/main/resources/application.yml):

```
micronaut:
  application:
    name: luiscmfunctionclient
aws:
  lambda:
    functions:
      helloworld: {}
    region: us-east-2
  accessKeyId: accessKeyGoesHere
  secretKey: secretKeyGoesHere
```

If I enter valid `accessKeyId` and `secretKey`, that application successfully invokes the expected Lambda.

```
~ $ http :8080/greet/Luis
HTTP/1.1 200 OK
connection: keep-alive
content-length: 39
content-type: application/json
date: Tue, 13 Jul 2021 19:21:53 GMT

{
    "greeting": "Hello Luis",
    "name": "Luis"
}

~ $ 
~ $ http :8080/greet/Jeff
HTTP/1.1 200 OK
connection: keep-alive
content-length: 39
content-type: application/json
date: Tue, 13 Jul 2021 19:21:56 GMT

{
    "greeting": "Hello Jeff",
    "name": "Jeff"
}
```

**Reply by jeffscottbrown** [reply]
The client is at [src/main/java/example/micronaut/HelloClient.java](https://github.com/jeffbrown/luiscmfunctionclient/blob/972346efe5b3fc8ef3f63e0f3331b522410c6966/src/main/java/example/micronaut/HelloClient.java):

```
package example.micronaut;

import io.micronaut.function.client.FunctionClient;

import javax.inject.Named;
import java.util.Map;

@FunctionClient
public interface HelloClient {
    @Named("helloworld")
    String greet(Map info);
}
```

**Comment by luiscm** [comment]
Dear Jeff

I was guessing that Micronaut would have a method to do it like I do in AWS
API.

In this example you make an app by calling a functionclient

I wanted a functionclient calling another functionclient, not as an
application.

Do I have a way to do it or not? Or it has to be the way the framework
works.

**Reply by jeffscottbrown** [reply]
Can you simply `@Inject` the client into your request handler?

**Reply by jeffscottbrown** [reply]
@luiscm Is injecting the client into your request handler sufficient to accomplish whatever it is you would like to accomplish?

**Comment by sdelamo** [comment]
I have written a documentation snippet here: https://micronaut-projects.github.io/micronaut-aws/latest/guide/#lambdafunctionclient
